pkg/lib/ldap/client.go

```diff
@@ -0,0 +1,109 @@
+package ldap
+
+import (
+	"bytes"
+	"errors"
+	"html/template"
+	"net/url"
+	"strings"
+
+	"github.com/authgear/authgear-server/pkg/lib/config"
+	"github.com/go-ldap/ldap/v3"
+)
+
+type UserProfile *ldap.Entry
+
+type LDAPClient struct {
+	Config config.LDAPServerConfig
+	conn   *ldap.Conn
+}
+
+func (c *LDAPClient) Connect() error {
+	conn, err := ldap.DialURL(c.Config.URL)
+	if err != nil {
+		return err
+	}
+
+	url, err := url.Parse(c.Config.URL)
+	if err != nil {
+		return err
+	}
+	if url.Scheme == "ldap" {
+		_ = conn.StartTLS(nil)
+	}
+
+	c.conn = conn
+	return nil
+}
+
+func (c *LDAPClient) Close() {
+	if c.conn != nil {
+		c.conn.Close()
+	}
+}
+
+func (c *LDAPClient) Bind(username string, password string) error {
+	if c.conn == nil {
+		return errors.New("ldap: not connected")
+	}
+	return c.conn.Bind(username, password)
+}
+
+func parseSearchFilter(filter string, username string) (string, error) {
+
+	trimedFilter := strings.TrimSpace(filter)
+
+	tmpl, err := template.New("search_filter").Parse(strings.Trim(trimedFilter, "\n"))
+	if err != nil {
+		return "", err
+	}
+	var buf bytes.Buffer
+	// make a string array of phone, username, email
+	err = tmpl.Execute(&buf, map[string]string{"Username": username})
```

Comment:
I found that this implementation allows user to make filter injection. Please escape the username before execute the template